File: 30 Days of Code/13 - Day 12 - Inheritance.py
# ========================
#       Information
# ========================

# Direct Link: https://www.hackerrank.com/challenges/30-inheritance/problem
# Difficulty: Easy
# Max Score: 30
# Language: Python

# ========================
#         Solution
# ========================

class Person:
    def __init__(self, first_name, last_name, id_number):
        self.first_name = first_name
        self.last_name = last_name
        self.id_number = id_number
    def print_person(self):
        '''Print Student Name and ID'''
        print("Name:", self.last_name + ",", self.first_name)
        print("ID:", self.id_number)

class Student(Person):
    '''Class Constructor'''
    # Parameters:
    # first_name - A string denoting the Person's first name.
    # last_name - A string denoting the Person's last name.
    # id - An integer denoting the Person's ID number.
    # scores - An array of integers denoting the Person's test scores.

    # Write your constructor here
    def __init__(self, first_name, last_name, id_number, scores):
        self.scores = scores
        super().__init__(first_name, last_name, id_number)

    # Function Name: calculate
    # Return: A character denoting the grade.

    # Write your function here
    def calculate(self):
        '''Calcuate Scores'''

        average = sum(self.scores) / len(self.scores)
        if 90 <= average <= 100:
            return 'O'
        elif 80 <= average <= 90:
            return 'E'
        elif 70 <= average <= 80:
            return 'A'
        elif 55 <= average <= 70:
            return 'P'
        elif 40 <= average <= 55:
            return 'D'
        else:
            return 'T'

LINE = input().split()
FIRST_NAME = LINE[0]
LAST_NAME = LINE[1]
ID_NUMBER = LINE[2]

NUM_SCORES = int(input()) # not needed for Python
SCORES = list( map(int, input().split()))

S = Student(FIRST_NAME, LAST_NAME, ID_NUMBER, SCORES)
S.print_person()

print("Grade:", S.calculate())

File: 30 Days of Code/C#/13 - Day 12 - Inheritance.cs
// ========================
//       Information
// ========================

// Direct Link: https://www.hackerrank.com/challenges/30-inheritance/problem
// Difficulty: Easy
// Max Score: 30
// Language: C#

// ========================
//         Solution
// ========================

using System;
using System.Linq;

class Person {
	protected string firstName;
	protected string lastName;
	protected int id;

	public Person() {}
	public Person(string firstName, string lastName, int identification) {
		this.firstName = firstName;
		this.lastName = lastName;
		this.id = identification;
	}
	public void printPerson() {
		Console.WriteLine("Name: " + lastName + ", " + firstName + "\nID: " + id);
	}
}

class Student: Person {
	private int[] testScores;

	// Class Constructor

	/* Parameters: 
    *  firstName - A string denoting the Person's first name.
    *  lastName - A string denoting the Person's last name.
    *  id - An integer denoting the Person's ID number.
    *  scores - An array of integers denoting the Person's test scores. */

	// Write your constructor here
	public Student(string firstName, string lastName, int id, int[] scores) {
		this.firstName = firstName;
		this.lastName = lastName;
		this.id = id;
		this.testScores = scores;
	}

    // Method Name: Calculate
    // Return: A character denoting the grade.

    // Write your method here
    public string Calculate() {
        var average = testScores.Average();
        var res = IsWithin(average, 90, 100) ? "O" : 
                    IsWithin(average, 80, 89) ? "E" :
                    IsWithin(average, 70, 79) ? "A" :
                    IsWithin(average, 50, 69) ? "P" :
                    IsWithin(average, 40, 54) ? "D" :
                    "T";
        return res;

    }

    private bool IsWithin(double value, int minimum, int maximum) {
		return value >= minimum && value <= maximum;
	}
}

class Solution {
	static void Main() {
		string[] inputs = Console.ReadLine().Split();
		string firstName = inputs[0];
		string lastName = inputs[1];
		int id = Convert.ToInt32(inputs[2]);
		int numScores = Convert.ToInt32(Console.ReadLine());
		inputs = Console.ReadLine().Split();
		int[] scores = new int[numScores];
		for (int i = 0; i < numScores; i++) {
			scores[i] = Convert.ToInt32(inputs[i]);
		}

		Student s = new Student(firstName, lastName, id, scores);
		s.printPerson();
		Console.WriteLine("Grade: " + s.Calculate() + "\n");
	}
}
